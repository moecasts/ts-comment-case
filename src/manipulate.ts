import * as ts from 'typescript';

/**
 * auto move following single line comment of the identifier before it
 */
export const moveTrailingCommentToLeading = (content: string) => {
  const src = ts.createSourceFile(
    'tmp.ts',
    content,
    ts.ScriptTarget.Latest,
    true,
  );

  let result = src.getFullText();
  let offset = 0;

  const consumedCommentRange = new Map<string, boolean>();

  const traverse = (node: ts.Node) => {
    const trailingCommentRanges = ts.getTrailingCommentRanges(
      src.getFullText(),
      node.getEnd(),
    );

    trailingCommentRanges?.forEach((range) => {
      const rangeKey = `${range.pos}-${range.end}`;
      if (consumedCommentRange.has(rangeKey)) {
        return;
      }

      const trailingComment = src.getFullText().slice(range.pos, range.end);

      let prefixText = src
        .getFullText()
        .slice(node.getFullStart(), node.getStart());

      if (node.getFullStart() !== 0) {
        const lastBreakLine = prefixText.lastIndexOf('\n');
        prefixText =
          lastBreakLine > -1
            ? prefixText.slice(lastBreakLine + 1, node.getStart())
            : '';
      }

      const indent = /^\s*/.test(prefixText)
        ? prefixText.match(/\n?([ \t]*)/)?.[1]
        : '';

      const leadingComment = trailingComment + '\n' + indent;

      // move trailing comment before the node
      result =
        result.slice(0, node.getStart() + offset) +
        leadingComment +
        node.getText() +
        result.slice(range.end + offset);

      offset += indent?.length || 0;

      consumedCommentRange.set(rangeKey, true);
    });

    ts.forEachChild(node, traverse);
  };

  traverse(src);

  return result;
};
