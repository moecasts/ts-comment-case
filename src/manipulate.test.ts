import { readFileSync } from 'fs';
import { join } from 'path';
import { describe, expect, test } from 'vitest';

import { moveTrailingCommentToLeading } from './manipulate';

const code2 = readFileSync(join(__dirname, '__materials__/code-2.ts'), 'utf-8');

describe('manipulate', () => {
  test('moveTrailingCommentToLeading', () => {
    const result = moveTrailingCommentToLeading(code2);

    expect(result).toMatchSnapshot();
  });
});
