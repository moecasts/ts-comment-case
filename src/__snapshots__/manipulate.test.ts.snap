// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`manipulate > moveTrailingCommentToLeading 1`] = `
"// @ts-nocheck
// This is a comment of var1
export const var1 = 1;

// This is a comment of var2
export const var2 = 2;

// this is a comment of Type
export type Type = string;

// This is a comment before Interface
export interface Interface {
  // This is a comment after Interface

  // this is a comment of type
  type: Type;
}

// This is a comment of function test
export function test() {
  // This is a comment before Interface
  interface InnerInterface {
    // This is a comment after Interface

    // this is a comment of type
    type: Type;
  }

  // this is a comment of InnerType
  type InnerType = string;

  // this is a comment of InnerType
  true && false;

  // eslint-disable-next-line
  // this is a comment of InnerType
  const a,
    b = 1 && false;

  type InnerTypeA = string;
  // this is a comment of InnerType
  type InnerTypeB = string;

  // This is a comment of console
  console.log('test');
}
"
`;
